cmd/util/common/checkpoint.go

```diff
@@ -0,0 +1,239 @@
+	// since a checkpoint file contains 500 tries, this allows us to find the block more efficiently
+	sealed, err := state.Sealed().Head()
+	if err != nil {
+		return nil, 0, flow.DummyStateCommitment, err
+	}
+	endHeight := sealed.Height
+
+	return GenerateProtocolSnapshotForCheckpointWithHeights(logger, state, headers, seals,
+		checkpointDir,
+		blocksToSkip,
+		endHeight,
+	)
+}
+
+// findLatestCheckpointFilePath finds the latest checkpoint file in the given directory
+// it returns the header file name of the latest checkpoint file
+func findLatestCheckpointFilePath(checkpointDir string) (string, error) {
+	_, last, err := wal.ListCheckpoints(checkpointDir)
+	if err != nil {
+		return "", fmt.Errorf("could not list checkpoints in directory %v: %w", checkpointDir, err)
+	}
+
+	fileName := wal.NumberToFilename(last)
+	if last < 0 {
+		fileName = "root.checkpoint"
+	}
+
+	checkpointFilePath := filepath.Join(checkpointDir, fileName)
+	return checkpointFilePath, nil
+}
+
+// GenerateProtocolSnapshotForCheckpointWithHeights does the same thing as GenerateProtocolSnapshotForCheckpoint
+// except that it allows the caller to specify the end height of the sealed block that we iterate backwards from.
+func GenerateProtocolSnapshotForCheckpointWithHeights(
+	logger zerolog.Logger,
+	state protocol.State,
+	headers storage.Headers,
+	seals storage.Seals,
+	checkpointDir string,
+	blocksToSkip uint,
+	endHeight uint64,
+) (protocol.Snapshot, uint64, flow.StateCommitment, error) {
+	startHeight := uint64(0)
+	// preventing startHeight from being negative
+	length := uint64(blocksToSkip+1) * 10000
+	if endHeight > length {
+		startHeight = endHeight - length
+	}
+
+	checkpointFilePath, err := findLatestCheckpointFilePath(checkpointDir)
+	if err != nil {
+		return nil, 0, flow.DummyStateCommitment, fmt.Errorf("could not find latest checkpoint file in directory %v: %w", checkpointDir, err)
+	}
+
+	log.Info().
+		Uint64("start_height", startHeight).
+		Uint64("end_height", endHeight).
+		Uint("blocksToSkip", blocksToSkip).
+		Msgf("generating protocol snapshot for checkpoint file %v", checkpointFilePath)
+	// find the height of the finalized block that produces the state commitment contained in the checkpoint file
+	sealedHeight, commit, err := FindHeightByCheckpoints(logger, headers, seals, checkpointFilePath, blocksToSkip, startHeight, endHeight)
+	if err != nil {
+		return nil, 0, flow.DummyStateCommitment, fmt.Errorf("could not find sealed height in range [%v:%v] (blocksToSkip: %v) by checkpoints: %w",
+			startHeight, endHeight, blocksToSkip,
+			err)
+	}
+
+	// find which finalized height seals the block with the given sealed height
+	finalizedHeight, err := findFinalizedHeightBySealedHeight(state, sealedHeight)
+	if err != nil {
+		return nil, 0, flow.DummyStateCommitment, fmt.Errorf("could not find finalized height for sealed height %v: %w", sealedHeight, err)
+	}
+
+	return state.AtHeight(finalizedHeight), sealedHeight, commit, nil
```

Comment:
this doesn't handle checking that the full sealing segment is included. Is that required for ENs?
https://github.com/onflow/flow-go/blob/542992558fce23c877e13011bccd6239207d5116/engine/access/rpc/backend/backend_network.go#L165

Reply:
Looks like the `getValidSnapshot` contains some necessary validation to the generated protocol snapshot, such as ensuring the finalized block and sealed block belong to the same epoch. 

However, these checks are placed in the access backend (engine/access/rpc/backend_network.go), I think it makes more sense to be placed in a package inside protocol state, so that it can be imported by other packages, such as util / admin. 

Any idea why we didn't do that in the first place? @jordanschalm

Reply:
Most likely the reason is that the protocol state needs to be able to return a snapshot for every incorporated block, which includes snapshots that `getValidSnapshot` considers invalid for dynamic bootstrapping. 

I'm happy to move the checks into the protocol package and agree that makes sense. It just needs to be differentiated from the existing snapshot API, something like `func GetDynamicBootstrapSnapshot(state, referenceSnapshot)`.